File: src/lib/routing/osmSurfaceLabel.ts
/** Honest OSM surface → asphalt / gravel / trail. Unknown keys stay raw. */
export type OsmSurfaceGroup = "asphalt" | "gravel" | "trail";

const ASPHALT = new Set([
  "asphalt",
  "paved",
  "concrete",
  "paving_stones",
  "sett",
  "chipseal",
  "concrete:plates",
  "concrete:lanes",
]);

const GRAVEL = new Set([
  "gravel",
  "compacted",
  "fine_gravel",
  "pebblestone",
  "unpaved",
  "ground",
]);

const TRAIL = new Set([
  "dirt",
  "earth",
  "grass",
  "path",
  "trail",
  "root",
  "mud",
  "sand",
  "wood",
  "winter_road",
]);

export function osmSurfaceGroup(raw: string | null | undefined): OsmSurfaceGroup | null {
  const key = (raw ?? "").trim().toLowerCase();
  if (!key) return null;
  if (ASPHALT.has(key)) return "asphalt";
  if (GRAVEL.has(key)) return "gravel";
  if (TRAIL.has(key)) return "trail";
  return null;
}

export function osmSurfaceLabel(
  raw: string | null | undefined,
  labels: { asphalt: string; gravel: string; trail: string }
): string {
  const key = (raw ?? "").trim();
  if (!key) return "";
  const group = osmSurfaceGroup(key);
  if (group === "asphalt") return labels.asphalt;
  if (group === "gravel") return labels.gravel;
  if (group === "trail") return labels.trail;
  return key.replaceAll("_", " ");
}

File: mobile/lib/domain/routing/osm_surface_label.dart
import '../../../l10n/app_localizations.dart';

enum OsmSurfaceGroup { asphalt, gravel, trail }

const _asphalt = {
  'asphalt',
  'paved',
  'concrete',
  'paving_stones',
  'sett',
  'chipseal',
  'concrete:plates',
  'concrete:lanes',
};

const _gravel = {
  'gravel',
  'compacted',
  'fine_gravel',
  'pebblestone',
  'unpaved',
  'ground',
};

const _trail = {
  'dirt',
  'earth',
  'grass',
  'path',
  'trail',
  'root',
  'mud',
  'sand',
  'wood',
  'winter_road',
};

OsmSurfaceGroup? osmSurfaceGroup(String? raw) {
  final key = (raw ?? '').trim().toLowerCase();
  if (key.isEmpty) return null;
  if (_asphalt.contains(key)) return OsmSurfaceGroup.asphalt;
  if (_gravel.contains(key)) return OsmSurfaceGroup.gravel;
  if (_trail.contains(key)) return OsmSurfaceGroup.trail;
  return null;
}

String osmSurfaceDisplay(String raw, AppLocalizations l10n) {
  final key = raw.trim();
  if (key.isEmpty) return '';
  final group = osmSurfaceGroup(key);
  if (group == OsmSurfaceGroup.asphalt) return l10n.filterSurfaceAsphalt;
  if (group == OsmSurfaceGroup.gravel) return l10n.filterSurfaceGravel;
  if (group == OsmSurfaceGroup.trail) return l10n.filterSurfaceTrail;
  return raw.replaceAll('_', ' ');
}
